# **Элементы теории погрешностей**

## **Источники погрешностей**

1. Математическая модель задачи является неточной (погрешность метода)

Погрешность возникает из-за того, что сам численный метод или математическая модель является лишь приближением к точному методу (например, дифференцирование). Кроме того, любая математическая модель или метод могут внести существенные погрешности, если в ней не учтены какие-то особенности рассматриваемой задачи. Модель может прекрасно работать в одних условиях и быть совершенно неприемлемой в других. Такую погрешность называют также методической. Она всегда имеет место, даже при абсолютно точных данных и абсолютно точных вычислениях. В большинстве случаев погрешность численного метода можно уменьшить до требуемого значения за счет изменения параметров метода (например, уменьшением шага дискретизации, или увеличением количества итераций).

2. Ошибки в исходных данных

Исходные данные задачи часто являются основным источником погрешностей. Ошибки такого типа неизбежны и проявляются в любых реальных задачах, поскольку любое измерение может быть проведено с только какой-то предельной точностью. Вместе с погрешностями, вносимыми математической моделью, их называют неустранимыми погрешностями, поскольку они не могут быть уменьшены ни до начала решения задачи, ни в процессе ее решения.

Следует стремиться к тому, чтобы все исходные данные были примерно одинаковой точности. Сильное уточнение одних исходных данных при наличии больших погрешностей в других не приводит к повышению точности конечных результатов. Если какие-то отдельные точки данных (измерения) явно ошибочные, их можно исключить из вычислений.

3. Вычислительные ошибки (ошибки округления)

Ошибки этого типа проявляются из-за дискретной (а не непрерывной) формы представления величин в компьютере. Вычислительные ошибки можно свести к минимуму продуманно организовывая алгоритмы.

## **Абсолютная и относительная погрешность**
Погрешностью вычислений  называется  отклонение  результата  вычисления от истинного значения числа. 
1. **Абсолютная погрешность** величины – разность между истинным и приближённым значениями этой величины. 
Пусть ***a*** - приближённое значение, ***x*** - точное значение некоторой величины, тогда абсолютная погрешность приближённого числа, определяется как

$$ \Delta_{a} = |x - a| $$

 Но поскольку ***х*** неизвестно, то и абсолютную погрешность мы узнать не можем! Чтобы разрешить парадокс, вводят предельную абсолютную погрешность
$$ \Delta_{a}^{*} $$ — такое значение, которое абсолютная погрешность заведомо

не превзойдет при данном способе измерений
$$ |x - a| \leq \Delta_{a}^{*} $$

 желательно возможно меньшее значение значение погрешности — это уменьшит длину интервала, содержащего искомое значение ***х*** и, следовательно, понизит неопределенность в знаниях об этой величине.


2. **Относительная погрешность** величины – отношение абсолютной погрешности к самой величине - 

$$ \delta_{a} =  \frac{\Delta_{a}}{|x|} $$ 

соответственно предельная относительная погрешность - 

$$ \delta_{a}^{*} =  \frac{\Delta_{a}^{*}}{|x|} $$ 

Относительная погрешность является величиной безразмерной,
т.е. не зависит от выбора системы единиц измерения, что позволяет
сравнивать качество измерений разнородных величин (бессмысленным является вопрос о том, что больше: 1 кг или 1 м,— но сравнение качества измерений массы и длины в терминах относительной погрешности вполне допустимо). 

## **Примеры**

### **Пример 1**
Для гравитационной постоянной предельная относительная погрешность

$$ \delta_{	\gamma}^{*} =  \frac{0.00085}{6.67259} = 1.27*10^{-4} $$ 


Для заряда электрона предельная относительная погрешность

$$ \delta_{	\gamma}^{*} =  \frac{0.00000049}{1.60217733} = 3.1*10^{-7} $$ 


Таким образом, в последнем случае относительная погрешность
оказывается на три порядка меньшей, т.е. заряд электрона определен
существенно точнее, чем гравитационная постоянная.


### **Пример 2**

С понятиями абсолютной и относительной погрешности связаны понятия верных и **значащих** цифр.
Если абсолютная погрешность приближенного числа **не превышает единицы последнего** (самого правого) **разряда его десятичной записи**, то цифры числа называют **верными** (или точными ).
По умолчанию десятичная запись приближенного числа должна содержать только верные цифры, и по записи числа сразу можно узнать предельную абсолютную погрешность, с которой оно известно.
Цифры, не являющиеся верными, называются сомнительными.

Даны приближенные числа а = 7.2, b = 2.40, c = 420,
d = 3.1∙10^3. Указать предельную абсолютную погрешность для каждого числа.
 
 Для числа а погрешность: 
 $ \Delta_{a}^{*} \leq 0.1  $; 

для числа b: $ \Delta_{b}^{*} \leq 0.01  $;

для числа с: $ \Delta_{c}^{*} \leq 10  $;

для числа d: $ \Delta_{d}^{*} \leq 100  $.


Итак, числа **а** и **b**, **с** и **d**, равные с точки зрения «обычной» математики, существенно различны в вычислительной математике: из абсолютной погрешности мы заключаем, что число **b** известно точнее, чем
число **а**, а число **с** — точнее, чем **d**. Нуль, стоящий справа в дробной части десятичного числа, важен, и им нельзя пренебрегать, если мы хотим составить верное суждение о точности числа.


### **Пример 3**

Найти предельные абсолютную и относительную погрешности
приближенного числа **x = 0.24**, все цифры которого по умолчанию верные.

Для нажождения абсолютной погрешности возьмём единицу последнего разряда  
 $$ \Delta_{x}^{*} \leq 0.01  $$ 

Соответственно относительная погрешность будет равна

 $$ \frac{0.01}{0.24} = 0.041666  $$

## **Индивидуальное задание**

а) Определить, какое равенство точнее.

б) Округлить сомнительные цифры числа, оставив верные знаки.
Определить абсолютную погрешность результата.

в) Найти предельные абсолютную и относительную погрешности
приближенного числа, все цифры которого по умолчанию верные.


### **1.** а) 14/17 = 0.824, $\sqrt{53}$ = 7.28; б) 23.3748, δ = 0.27%; в) 0.645. 
### **2.** а) 7/3 = 2.33, $\sqrt{58}$ = 7.62; б) 13.5726 $\pm$ 0.0072; в) 4.8556. 
### **3.** а) 27/31 = 0.871, $\sqrt{42}$ = 6.48; б) 0.088748, δ = 0.56%; в) 71.385. 
### **4.** а) 23/9 = 2.56, $\sqrt{87}$ = 9.33; б) 4.57633 $\pm$ 0.00042; в) 6.8346. 
### **5.** а) 6/7 = 0.857, $\sqrt{41}$ = 6.40; б) 46.7843, δ =  0.32%; в) 7.38. 
### **6.** а) 12/7 = 1.71, $\sqrt{47}$ = 6.86; б) 0.38725 $\pm$ 0.00112; в) 0.00646. 
### **7.** а) 21/13 = 1.62, $\sqrt{63}$ = 7.94; б) 45.7832, δ =  0.18%; в) 3.6765. 
### **8.** а) 16/7 = 2.29, $\sqrt{11}$ = 3.32; б) 0.75244 $\pm$ 0.00013; в) 5.374. 
### **9.** а) 18/7 = 2.57, $\sqrt{22}$ = 4.69; б) 46.453, δ = 0.15%; в) 6.125.
### **10.** а) 17/9 = 1.89, $\sqrt{17}$ = 4.12; б) 0.66385 $\pm$ 0.00042; в) 24.6. 
### **11.** а) 51/11 = 4.64, $\sqrt{35}$ = 5.92; б) 0.66385, δ = 0.34%; в) 0.543.
### **12.** а) 19/12 = 1.58, $\sqrt{12}$ = 3.46; б) 4.88445 $\pm$ 0.00052; в) 4.633. 
### **13.** а) 13/7 = 1.857, $\sqrt{7}$ = 2.65; б) 2.8867, δ = 0.43%; в) 63.749.
### **14.** а) 49/13 = 3.77, $\sqrt{14}$ = 3.74; б) 5.6483 $\pm$ 0.0017; в) 0.00858. 
### **15.** а) 5/3 = 1.667, $\sqrt{38}$ = 6.16; б) 3.7542, δ = 0.32%; в) 0.389.
### **16.** а) 17/11 = 1.545, $\sqrt{18}$ = 4.243; б) 0.8647 $\pm$ 0.0013; в) 0.864. 
### **17.** а) 7/22 = 0.318, $\sqrt{13}$ = 3.61; б) 0.3944, δ = 0.15%; в) 21.7.
### **18.** а) 13/17 = 0.765, $\sqrt{31}$ = 5.57; б) 3.6878 $\pm$ 0.0013; в) 8.74. 
### **19.** а) 50/19 = 2.63, $\sqrt{27}$ = 5.20; б) 0.85638, δ = 0.22%; в) 231.57.
### **20.** а) 21/29 = 0.724, $\sqrt{44}$ = 6.63; б) 13.6853 $\pm$ 0.0023; в) 2.16. 
### **21.** а) 17/19 = 0.895, $\sqrt{52}$ = 7.21; б) 7.521, δ = 0.12%; в) 0.5748.
### **22.** а) 6/11 = 0.545, $\sqrt{83}$ = 9.11; б) 3.7832 $\pm$ 0.0043; в) 2.678. 
### **23.** а) 16/19 = 0.842, $\sqrt{55}$ = 7.416; б) 17.356, δ = 0.11%; в) 0.5718.
### **24.** а) 23/15 = 1.53, $\sqrt{98}$ = 9.899; б) 8.7432 $\pm$ 0.0023; в) 0.578. 
### **25.** а) 2/21 = 0.095, $\sqrt{22}$ = 4.69; б) 24.5641, δ = 0.09%; в) 4.478.

### **Разбор. Пример 25.**


#### **а)** Определить, какое равенство точнее. 
**2/21 = 0.095, $\sqrt{22}$ = 4.69**

In [ ]:
a = 2/21
b = 0.095

c = 22 ** 0.5
d = 4.69

print(a," = ",b,"; ",c," = ",d)

0.09523809523809523  =  0.095 ;  4.69041575982343  =  4.69


In [ ]:
abs1 = abs(a-b)
abs2 = abs(c-d)
print(abs1, abs2)

0.0002380952380952317 0.00041575982342934026


In [ ]:
rel1 = round(abs1/a, 5)
rel2 = round(abs2/c, 5)

print(rel1,rel2)

0.0025 9e-05


используя функцию **float** можно вывести значение c 6 знаками после запятой и без экспоненциальной записи:

In [ ]:
'{:f}'.format(rel2)

'0.000090'

Так как предельная относительная погрешность второго равенства меньше (**0.0025 > 0.00009**), оно точнее.

#### **б)** Округлить сомнительные цифры числа, оставив верные знаки. Определить абсолютную погрешность результата.
**24.5641, δ = 0.09%**

In [ ]:
a = 25.5641
rel1 = 0.09 / 100 # так как в процентах
abs1 = rel1 * a 
round(abs1, 5)

0.02301

Абсолютная погрешность результата составляет: **0.02301**
Число с верными знаками равно **25.5** так как, абсолютная погрешность не превышает единицы последнего разряда. Если оставить 4 знака, то погрешность уже превышает единицу последнего разряда. 


#### **в)** Найти предельные абсолютную и относительную погрешности приближенного числа, все цифры которого по умолчанию верные.
**4.478**

In [ ]:
a = 4.478

Если все цифры числа верные(значащие), значит предельная абсолютная погрешность не превышает единицу последнего разряда. Следует:

In [ ]:
abs1 = 0.001

Предельная относительная погрешность:

In [ ]:
rel1 = abs1 / abs(a)
round(rel1,5)

0.00022

Предельная абсолютная погрешность: **0.001**

Предельная относительная погрешность: **0.00022**

## **Полезные ссылки:**

1. [Погрешность измерения на вики](https://ru.wikipedia.org/wiki/%D0%9F%D0%BE%D0%B3%D1%80%D0%B5%D1%88%D0%BD%D0%BE%D1%81%D1%82%D1%8C_%D0%B8%D0%B7%D0%BC%D0%B5%D1%80%D0%B5%D0%BD%D0%B8%D1%8F)
2. [Подробнее о функции ***format***](https://pyformat.info/)
3. [Зенков А.В. Численные методы. (*Глава 1. Погрешности - стр. 9-17*)](https://urait.ru/viewer/chislennye-metody-452829?embded=1#page/9)